CLOUDSTACK-9074: Support shared networking in NiciraNVP Plugin

nvazquez opened the pull request:
JIRA TICKET: 
https://issues.apache.org/jira/browse/CLOUDSTACK-9074 

Design Document:
https://cwiki.apache.org/confluence/display/CLOUDSTACK/Support+shared+networking+in+NiciraNVP+Plugin

remibergsma commented:
LGTM, based on a set of tests that I run on this branch (which I rebased myself first). Screenshot is from an experimental Jenkins job (that runs the same tests I did manually until now).

These test may not cover your change, all they do is show you didn't break them. Someone else needs to review the code.

![1094_test_results](https://cloud.githubusercontent.com/assets/1630096/11296170/d08de85e-8f70-11e5-8d0f-3ae7a735afa1.png)

nvazquez commented:
Thanks @remibergsma !

remibergsma commented:
@nvazquez Ai, there is a conflict now. Could you please have a look?

nvazquez commented:
@remibergsma thanks Remi, I rebased to master branch and resolved a few conflicts

remibergsma commented:
@miguelaferreira @wilderrodrigues Can one of you review this? Once done, I can run the integration tests again.Thanks!

miguelaferreira commented on api/src/com/cloud/network/Networks.java:
```diff
@@ -251,6 +252,10 @@ public static URI fromString(String candidate) {
                 }
+                if (UuidUtils.validateUUID(candidate)){
+                    //Supporting lswitch uuid as vlan id: set broadcast_uri = null and add row on nicira_nvp_router_map with logicalrouter_uuid = candidate
+                    return null;
```
I've read the comment in the line above a few times, and still cannot understand the purpose of returning null on an invalid UUID (i.e. you previously caught a number format exception).

nvazquez replied:
The idea is to support creating Shared Networks over NSX Managed Network Offering, so there will be this two cases (for creation):
* Providing numerical VLAN_ID:
 * broadcastdomain_type = Lswitch
 * broadcast_uri = "vlan://VLAN_ID"
* Providing logical router UUID as VLAN_ID:
 * broadcastdomain_type = Lswitch
 * broadcast_uri = NULL
 * add entry on nicira_nvp_router_map with network id and logical router's UUID

For second case, I included the code in the catch block as I asumed that VLAN_ID provided is not numerical (as long parsing fail).

miguelaferreira replied:
Ok, I understand the intention now. However, you could make that intention clear in the code. 

You use exception handling to deal with expected behaviour because you implement the expected behaviour in the `catch`block. Given the state-of-the-practive in this project I would say that is normal, and I accept what you did.

Still, if you would like to make it better, you could create a method `private boolean isVlanId(String candidate)` that has the `try...catch` in there (with the same construction you have now) but returns `true`or `false`. Then in the `fromString` method you can call the new method, and instead of having your logic in the `catch` block you will have it in a `else` block. This will make clear that the code is not handling an exceptional (error) case.
It would look a bit like this:
```java
(...)

private boolean isVlanId(String candidate) {
  try {
    Long.parseLong(candidate);
    return true;
  } catch (NumberFormatException nfe) {
    return false;
  }
}

(...)

if(isVlanId(candidate)) {
  return Vlan.toUri(candidate);
} else {
  (...)
}

(...)
```

In addition, instead of returning `null` (because that is a very bad practice, and opens up an entire category of errors, a.k.a. null dereferencing) you could return Optional<URI> and then in the method that calls `fromString` you need to handle the case of an existing URI or a non-existing URI. Optional has an [API](https://code.google.com/p/guava-libraries/wiki/UsingAndAvoidingNullExplained) for that.
The idea behind this recommendation is that we express the fact that we might have an URI or not, in a safer and more meaningful way that returning `null`.

miguelaferreira replied:
One other thing, isn't there a method that validates a UUID? 

Maybe I'm missing something, but wound't `UuidUtils.validateUUID(candidate)` tell you if candidate is a UUID or not?
If so, why not use that instead of the construction with `try { Long.parse(...); } catch (NumberFormatException nfe) { (...) }`?

miguelaferreira commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -671,8 +672,19 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
+
+                        if (predefined instanceof NetworkVO){
+                            try{
+                                NetworkGuruAdditionalFunctions functions = (NetworkGuruAdditionalFunctions) guru;
```
This would best be protected with a `instanceof` check.

miguelaferreira commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -671,8 +672,19 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
+                                functions.finalizeNetworkDesign(networkPersisted.getId(), ((NetworkVO)predefined).getVlanIdAsUUID());
+                            }
+                            catch (Exception e){
+                                //Couldn't cast to NetworkGuruAdditionalFunctions, continue
```
The code should not catch exception because it is simply too broad.
Caught exceptions cannot be silently ignored (i.e. having an empty `catch` block).

miguelaferreira commented on api/src/com/cloud/network/guru/NetworkGuruAdditionalFunctions.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.util.Map;
+
+public interface NetworkGuruAdditionalFunctions {
```
Several problems with this interface. An interface called functions seems like it is meant o `patch up` another interface that is lacking functionality. Secondly, it looks very much NSX specific while the name and package seem quite generic.

nvazquez replied:
I agree. This was the "best" solution I found, maybe for my lack CS architecture knowledge. This was the problem that I faced:
I needed to call this functions from NetworkOrchestrator. However, from NetworkOrchestrator I could access NetworkGuru interface, so I decided not to include this methods in NetworkGuru interface, implementing them blank in every guru except in NiciraNvpGuestNetworkGuru. I decided to create this interface, only implemented by NiciraNvpGuestNetworkGuru just to avoid doing that.

In NetworkOrchestrator, line 682 I made a call to one of this functions. How can I improve this solution?

miguelaferreira replied:
Wound't it be possible to call these methods inside the `GuestNetworkGuru` methods?
I mean, have `NiciraNvpGuestNetworkGuru` call `finalizeNetworkDesign` and `listAdditionalNicParams` within the methods that are already defined in NetworkGuru?

miguelaferreira commented on api/src/org/apache/cloudstack/api/response/NicResponse.java:
```diff
@@ -102,6 +102,14 @@
 
+    @SerializedName("nsxlogicalswitch")
+    @Param(description = "Id of the NSX Logical Switch (if NSX based), null otherwise", since="4.6.0")
+    private String nsxlogicalswitch;
```
variable naming convention not being followed here. It should be `nsxLogicalSwitch`

miguelaferreira commented on api/src/org/apache/cloudstack/api/response/NicResponse.java:
```diff
@@ -102,6 +102,14 @@
+
+    @SerializedName("nsxlogicalswitchport")
+    @Param(description = "Id of the NSX Logical Switch Port (if NSX based), null otherwise", since="4.6.0")
+    private String nsxlogicalswitchport;
```
variable naming convention not being followed here.

miguelaferreira commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -671,8 +672,19 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
                         vo.setStrechedL2Network(offering.getSupportsStrechedL2());
-                        networks.add(_networksDao.persist(vo, vo.getGuestType() == Network.GuestType.Isolated,
-                                finalizeServicesAndProvidersForNetwork(offering, plan.getPhysicalNetworkId())));
+                        NetworkVO networkPersisted = _networksDao.persist(vo, vo.getGuestType() == Network.GuestType.Isolated,
```
This block adds more code to an already oversized method. This would best be extracted as a new method or even a new object. This way it would be easier to read and unit-test.

miguelaferreira commented on plugins/network-elements/nicira-nvp/src/main/java/com/cloud/network/element/NiciraNvpElement.java:
```diff
@@ -239,8 +244,51 @@ public boolean implement(Network network, NetworkOffering offering, DeployDestin
 
-        // Implement SourceNat immediately as we have al the info already
-        if (networkModel.isProviderSupportServiceInNetwork(network.getId(), Service.SourceNat, Provider.NiciraNvp)) {
+        if (network.getGuestType().equals(GuestType.Shared)){
```
This block (and the next) add to an already large method. It would be best to refactor it to new methods (or a new object altogether).

nvazquez replied:
Ok, done. Is it ok like this?

miguelaferreira replied:
Yes! Much better. Thanks!

miguelaferreira commented on plugins/network-elements/nicira-nvp/src/main/java/com/cloud/network/guru/NiciraNvpGuestNetworkGuru.java:
```diff
@@ -102,11 +112,11 @@ public NiciraNvpGuestNetworkGuru() {
         // This guru handles only Guest Isolated network that supports Source nat service
-        if (networkType == NetworkType.Advanced && isMyTrafficType(offering.getTrafficType()) && offering.getGuestType() == Network.GuestType.Isolated
+        if (networkType == NetworkType.Advanced && isMyTrafficType(offering.getTrafficType())
+                && (offering.getGuestType() == Network.GuestType.Isolated || offering.getGuestType() == Network.GuestType.Shared)
```
Is this check really needed? I mean `Isolated` and `Shared` are the [only two types of guest networks](https://github.com/apache/cloudstack/blob/95ae7963d5652595d62d6ace39de2eec9670413c/api/src/com/cloud/network/Network.java#L41).
If you would like to keep the check for sake of being future proof (that is, new guest types being added), then refactoring it to a method with a meaningful name (e.g. supportedGuestTypes(...)) would help readability.

miguelaferreira commented on plugins/network-elements/nicira-nvp/src/main/java/com/cloud/network/resource/wrapper/NiciraNvpConfigureSharedNetworkUuidCommandWrapper.java:
```diff
@@ -0,0 +1,110 @@
+                    if (lsp != null){
+                        niciraNvpApi.deleteLogicalSwitchPort(logicalSwitchUuid, lsp.getUuid());
+                    }
+                } catch (NiciraNvpApiException ex) {
```
Caught exceptions should at least be logged.

nvazquez replied:
Ok, I added logs on these <code>catch</code> blocks

miguelaferreira replied:
This is indeed better. However, now you are logging it as an info message, and then making a new exception.
If you log something, and then throw an exception, I would say the log level should be warn because you are delegating the handling of the exception to the callers.
In addition, when you make a new exception out of an existing exception you should include the original exception as the cause of the new exception.
In this case you are using two exception as the cause `e and ex`. 

Let us think of a way of doing this better:
* The first exception `ex` is being handled here. So I would say, first thing is to log it as an error (`s_logger.error("failed to do something", ex)`). This way we explain what failed and present the cause in the logs.
* Then we handle that exception (`ex`) and we might get a second one (`e`). 
* This second one we don't want to handle here, and we will bubble it up wrapped in a new exception. So then we log `e` as a warning (`s_logger.warn("failed to do something else, namely handing the previous excpetion", e)`), and then we wrap it in  a new exception and throw it (`throw new NiciraNvpApiException("probably the same message we set in the warn log", e`).

What do you think? Does this make sense to you?

miguelaferreira commented on plugins/network-elements/nicira-nvp/src/main/java/com/cloud/network/resource/wrapper/NiciraNvpConfigureSharedNetworkUuidCommandWrapper.java:
```diff
@@ -0,0 +1,110 @@
+                       logicalRouterUuid + ". Network gateway " + portIpAddress +
+                       " (" + e.getMessage() + ")";
+                throw new NiciraNvpApiException(msg);
+            } catch (NiciraNvpApiException e1) {
```
Caught exceptions should at least be logged.

miguelaferreira commented on plugins/network-elements/nicira-nvp/src/main/java/com/cloud/network/resource/wrapper/NiciraNvpConfigureSharedNetworkVlanIdCommandWrapper.java:
```diff
@@ -0,0 +1,93 @@
+            catch (NiciraNvpApiException e){
+                try {
+                    niciraNvpApi.deleteLogicalSwitchPort(logicalSwitchUuid, lsp.getUuid());
+                } catch (NiciraNvpApiException e1) {
```
Caught exceptions should at least be logged.

miguelaferreira commented on plugins/network-elements/nicira-nvp/src/main/java/com/cloud/network/resource/wrapper/NiciraNvpConfigureSharedNetworkVlanIdCommandWrapper.java:
```diff
@@ -0,0 +1,93 @@
+        }
+        catch (NiciraNvpApiException e){
+            try {
+                throw e;
```
You except an exception when throwing an exception?

miguelaferreira commented on setup/db/db/schema-460to470.sql:
```diff
@@ -18,3 +18,5 @@
 -- Schema upgrade from 4.6.0 to 4.7.0;
 --;
+
+ALTER TABLE `cloud`.`nicira_nvp_router_map` DROP INDEX `logicalrouter_uuid` ;
```
What's the purpose behind dropping the index?

miguelaferreira replied:
This is about not being unique anymore, right?

nvazquez replied:
Right. As many networks can now connect to the same logical router

miguelaferreira commented on setup/db/db/schema-461to470.sql:
```diff
@@ -18,3 +18,5 @@
 -- Schema upgrade from 4.6.1 to 4.7.0;
 --;
+
+ALTER TABLE `cloud`.`nicira_nvp_router_map` DROP INDEX `logicalrouter_uuid` ;
```
What's the purpose behind dropping the index?

miguelaferreira replied:
This is about not being unique anymore, right?

nvazquez replied:
Right. As many networks can now connect to the same logical router

miguelaferreira commented:
@nvazquez I've just read through you code and I made some comments that I would be happy to discuss with you.

Nice to see effort being put in unit testing the PR.

nvazquez commented:
@miguelaferreira thanks for reviewing! Sure, would be great to discuss them

nvazquez commented:
@miguelaferreira besides fixing these issues and writing marvin test, what else would be needed to include this feature in version 4.7? Would it be possible, or could be included in later versions?

miguelaferreira commented:
@nvazquez I have one more request for you, and (if you agree, of course) that should be the end of it on a code level. With the marvin test then this would be ready for 4.7, but be aware that feature freeze is planned for December 7, so we need to work fast.
However, if we do not make that deadline we can always include this feature in 4.8 (as you have probably seen, we are releaseing more often these days :smile:)

I have to say I've never really liked reviewing code for other people, but it has been a pleasure to do it with you :+1:.

nvazquez commented:
Thanks @miguelaferreira I really appreciate that! Thanks a lot for your help!
I'll try to make it before the deadline. I'm not that into Python but I'll do my best with Marvin tests

miguelaferreira commented:
I'm working on adding a new test case to the test file we already have, but unfortunately I don't know how to cover your change. What I do know is how marvin works, and how a test case should look like, so please do ping me (here, or by mail) so that we can setup a chat somewhere and I'll be happy to help you.

nvazquez commented:
Sure, I couldn't find you mail address, mine is in my profile

remibergsma commented:
@nvazquez @miguelaferreira Hi Guys, do you think you can finish this PR on Monday so we can include it in 4.7? Would be nice :-) Let me know if you need help.

nvazquez commented:
Hi @remibergsma @miguelaferreira 

I've been working in marvin test file to cover this new features, those are test03 y test04 in <code>test/integration/smoke/test_nicira_controller.py</code>

I attach my test results:
```
$ cat /tmp/MarvinLogs/test_nicira_controller_1Z39R8/results.txt
test_01_nicira_controller (integration.smoke.test_nicira_controller.TestNiciraContoller) ... === TestName: test_01_nicira_controller | Status : SUCCESS ===
ok
Nicira clusters will redirect clients (in this case ACS) to the master node. ... === TestName: test_02_nicira_controller_redirect | Status : SUCCESS ===
ok
Shared Networks Support ... === TestName: test_03_nicira_shared_networks_numerical_vlanid | Status : SUCCESS ===
ok
# Shared Networks Support ... === TestName: test_04_nicira_shared_networks_lrouter_uuid_vlan_id | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 4 tests in 1415.297s

OK

```

miguelaferreira commented:
I'm finishing my own test, and will test this PR asap.

miguelaferreira commented:
@nvazquez I've been able to verify that the existing tests did not break with your PR, however I can't seem to get an NSX configuration for running the two additional test you added.

Here's my network config:
```
start IP = 192.168.23.2
end IP   = 192.168.23.20
netmask  = 255.255.255.0
gateway  = 192.168.23.1
```

The VLAN config
```
VLAN            = 50
VLAN_UUID = 2eea1553-837e-4a3e-bfbb-872c4cb2a49a
```

The VLAN_UUID is the UUID of a LRouter I've created in NSX. That LRouter is connected to an LSwitch.
The pictures bellow show the LRouter and LSwitch.

<img width="963" alt="screen shot 2015-12-06 at 11 50 07" src="https://cloud.githubusercontent.com/assets/4670993/11612955/a094b276-9c0f-11e5-9774-657ae8b21f9f.png">

<img width="1118" alt="screen shot 2015-12-06 at 11 50 15" src="https://cloud.githubusercontent.com/assets/4670993/11612956/a64278ca-9c0f-11e5-8206-a8ab3afd0d96.png">

<img width="1121" alt="screen shot 2015-12-06 at 11 50 25" src="https://cloud.githubusercontent.com/assets/4670993/11612958/abe15d32-9c0f-11e5-8fd8-60f0d2f9bc59.png">

<img width="1014" alt="screen shot 2015-12-06 at 11 50 32" src="https://cloud.githubusercontent.com/assets/4670993/11612959/afd87a24-9c0f-11e5-92f6-08dfe0491c5a.png">

If you could help me fix my config, I'll test your PR today.

serg38 commented:
Since you are testing L3 connectivity existing switch patch to router shouldn't overlap an ip range used by the new subnet. Cidr #1 should work fine . You might also try disconnect existing switch from router to see if test passes. For test #4 we tested we lrouter patched over to l3 gateway without existing switch . For test# 3 you need to have l2 gateway to match the one specified for nvp device. The issue we observed was if any of the test fail for any reason the cleanup on nsx side doesn't work so a manual cleanup might require to remove existing router port left over from previous tests that has overlapping cidr

miguelaferreira commented:
@serg38 I will use CIDR 1 and disconnect the switch from the router, to se if that works. I haven't' seen any think left behind after testing. That is, the clean up seems to be working fine for me.

remibergsma commented:
Guys, I cannot build a cloud from this branch any more. The build itself works, but the database deployment gives an SQL error: `Can't DROP 'logicalrouter_uuid'; check that column/key exists`

Details:

```
[INFO] --- exec-maven-plugin:1.2.1:java (create-schema) @ cloud-developer ---
log4j:WARN No appenders could be found for logger (org.springframework.core.env.StandardEnvironment).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
========> WARNING: Provided file does not exist: /data/git/cs1/cloudstack/developer/../utils/conf/db.properties.override
========> WARNING: Provided file does not exist: /data/git/cs1/cloudstack/developer/developer-prefill.sql.override
========> Initializing database=cloud with host=localhost port=3306 username=cloud password=cloud
============> Running query: drop database if exists `cloud`
============> Running query: create database `cloud`
============> Running query: GRANT ALL ON cloud.* to 'cloud'@`localhost` identified by 'cloud'
============> Running query: GRANT ALL ON cloud.* to 'cloud'@`%` identified by 'cloud'
========> Initializing database=cloud_usage with host=localhost port=3306 username=cloud password=cloud
============> Running query: drop database if exists `cloud_usage`
============> Running query: create database `cloud_usage`
============> Running query: GRANT ALL ON cloud_usage.* to 'cloud'@`localhost` identified by 'cloud'
============> Running query: GRANT ALL ON cloud_usage.* to 'cloud'@`%` identified by 'cloud'
========> Processing SQL file at /data/git/cs1/cloudstack/developer/target/db/create-schema.sql
========> Processing SQL file at /data/git/cs1/cloudstack/developer/target/db/create-schema-premium.sql
========> Processing SQL file at /data/git/cs1/cloudstack/developer/target/db/templates.sql
========> Processing SQL file at /data/git/cs1/cloudstack/developer/developer-prefill.sql
========> Processing upgrade: com.cloud.upgrade.DatabaseUpgradeChecker
[WARNING]
java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:297)
        at java.lang.Thread.run(Thread.java:745)
Caused by: com.cloud.utils.exception.CloudRuntimeException: Unable to execute upgrade script: /data/git/cs1/cloudstack/developer/target/db/db/schema-410to420-cleanup.sql
        at com.cloud.upgrade.DatabaseUpgradeChecker.runScript(DatabaseUpgradeChecker.java:290)
        at com.cloud.upgrade.DatabaseUpgradeChecker.upgrade(DatabaseUpgradeChecker.java:425)
        at com.cloud.upgrade.DatabaseUpgradeChecker.check(DatabaseUpgradeChecker.java:474)
        at com.cloud.upgrade.DatabaseCreator.main(DatabaseCreator.java:217)
        ... 6 more
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Can't DROP 'logicalrouter_uuid'; check that column/key exists
        at com.cloud.utils.db.ScriptRunner.runScript(ScriptRunner.java:185)
        at com.cloud.utils.db.ScriptRunner.runScript(ScriptRunner.java:87)
        at com.cloud.upgrade.DatabaseUpgradeChecker.runScript(DatabaseUpgradeChecker.java:281)
        ... 9 more
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 19.639s (Wall Clock)
[INFO] Finished at: Sun Dec 06 09:07:53 GMT 2015
[INFO] Final Memory: 43M/259M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:java (create-schema) on project cloud-developer: An exception occured while executing the Java class. null: InvocationTargetException: Unable to execute upgrade script: /data/git/cs1/cloudstack/developer/target/db/db/schema-410to420-cleanup.sql: Can't DROP 'logicalrouter_uuid'; check that column/key exists -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

Not sure if it is related. Tried cleaning everything (VMs, git repo, storage) but same result.

serg38 commented:
This index comes from schema-40to410.sql:

CREATE TABLE `cloud`.`nicira_nvp_router_map` (

  `id` bigint unsigned NOT NULL AUTO_INCREMENT COMMENT 'id',

  `logicalrouter_uuid` varchar(255) NOT NULL UNIQUE COMMENT 'nicira uuid of logical router',

  `network_id` bigint unsigned NOT NULL UNIQUE COMMENT 'cloudstack id of the network',

  PRIMARY KEY (`id`),

  CONSTRAINT `fk_nicira_nvp_router_map__network_id` FOREIGN KEY (`network_id`) REFERENCES `networks`(`id`) ON DELETE CASCADE

) ENGINE=InnoDB DEFAULT CHARSET=utf8; 







There might be variation how UNIQUE attribute is enforced in different versions of MySQL.

In 5.5 it generate unique index  `logicalrouter_uuid`

To take it out the correct statement is

ALTER TABLE `cloud`.`nicira_nvp_router_map` 
DROP INDEX `logicalrouter_uuid` ;

Can you check if schema-40to410.sql got applied in your environment? What version of MySQL do you use?

The drop index logicalrouter_uuid should come from schema-470to471.sql (or later upgrade script)but not from schema-410to420-cleanup.sql.



From:  Remi Bergsma <notifications@github.com>
Reply-To:  apache/cloudstack <reply@reply.github.com>
Date:  Sunday, December 6, 2015 at 11:59 AM
To:  apache/cloudstack <cloudstack@noreply.github.com>
Cc:  Sergey Levitskiy <serg38l@hotmail.com>
Subject:  Re: [cloudstack] CLOUDSTACK-9074: Support shared networking in NiciraNVP Plugin (#1094)


Guys, I cannot build a cloud from this branch any more. The build itself works, but the database deployment gives an SQL error: Can't DROP 'logicalrouter_uuid'; check that column/key exists

Details:
[INFO] --- exec-maven-plugin:1.2.1:java (create-schema) @ cloud-developer ---
log4j:WARN No appenders could be found for logger (org.springframework.core.env.StandardEnvironment).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
========> WARNING: Provided file does not exist: /data/git/cs1/cloudstack/developer/../utils/conf/db.properties.override
========> WARNING: Provided file does not exist: /data/git/cs1/cloudstack/developer/developer-prefill.sql.override
========> Initializing database=cloud with host=localhost port=3306 username=cloud password=cloud
============> Running query: drop database if exists `cloud`
============> Running query: create database `cloud`
============> Running query: GRANT ALL ON cloud.* to 'cloud'@`localhost` identified by 'cloud'
============> Running query: GRANT ALL ON cloud.* to 'cloud'@`%` identified by 'cloud'
========> Initializing database=cloud_usage with host=localhost port=3306 username=cloud password=cloud
============> Running query: drop database if exists `cloud_usage`
============> Running query: create database `cloud_usage`
============> Running query: GRANT ALL ON cloud_usage.* to 'cloud'@`localhost` identified by 'cloud'
============> Running query: GRANT ALL ON cloud_usage.* to 'cloud'@`%` identified by 'cloud'
========> Processing SQL file at /data/git/cs1/cloudstack/developer/target/db/create-schema.sql
========> Processing SQL file at /data/git/cs1/cloudstack/developer/target/db/create-schema-premium.sql
========> Processing SQL file at /data/git/cs1/cloudstack/developer/target/db/templates.sql
========> Processing SQL file at /data/git/cs1/cloudstack/developer/developer-prefill.sql
========> Processing upgrade: com.cloud.upgrade.DatabaseUpgradeChecker
[WARNING]
java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:297)
        at java.lang.Thread.run(Thread.java:745)
Caused by: com.cloud.utils.exception.CloudRuntimeException: Unable to execute upgrade script: /data/git/cs1/cloudstack/developer/target/db/db/schema-410to420-cleanup.sql
        at com.cloud.upgrade.DatabaseUpgradeChecker.runScript(DatabaseUpgradeChecker.java:290)
        at com.cloud.upgrade.DatabaseUpgradeChecker.upgrade(DatabaseUpgradeChecker.java:425)
        at com.cloud.upgrade.DatabaseUpgradeChecker.check(DatabaseUpgradeChecker.java:474)
        at com.cloud.upgrade.DatabaseCreator.main(DatabaseCreator.java:217)
        ... 6 more
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Can't DROP 'logicalrouter_uuid'; check that column/key exists
        at com.cloud.utils.db.ScriptRunner.runScript(ScriptRunner.java:185)
        at com.cloud.utils.db.ScriptRunner.runScript(ScriptRunner.java:87)
        at com.cloud.upgrade.DatabaseUpgradeChecker.runScript(DatabaseUpgradeChecker.java:281)
        ... 9 more
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 19.639s (Wall Clock)
[INFO] Finished at: Sun Dec 06 09:07:53 GMT 2015
[INFO] Final Memory: 43M/259M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:java (create-schema) on project cloud-developer: An exception occured while executing the Java class. null: InvocationTargetException: Unable to execute upgrade script: /data/git/cs1/cloudstack/developer/target/db/db/schema-410to420-cleanup.sql: Can't DROP 'logicalrouter_uuid'; check that column/key exists -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Not sure if it is related. Tried cleaning everything (VMs, git repo, storage) but same result.

—
Reply to this email directly or view it on GitHub.

nvazquez commented:
@remibergsma my mistake, I meant to modify <code>schema-460to470.sql</code> and <code>schema-461to470.sql</code> but not <code>setup/db/db/schema-410to420-cleanup.sql</code> as I did.
I'll push some other changes and also fix this.

remibergsma commented:
@nvazquez OK, check! No problem. Just let me know when you're ready and I'll run some more tests.

nvazquez commented:
ping @remibergsma would be great to test this new functionalities

nvazquez commented:
@miguelaferreira configuration needed is very complicated, I could not be able to include it on <code>test/integration/plugins/test_nicira_controller.py</code>, tests added assume that configuration is ok.

I attach 2 documents that could help you with transport zone and L2 Gateway configuration:
[NSX-MH-DesignGuidev4.2_Final.pdf](https://github.com/apache/cloudstack/files/57305/NSX-MH-DesignGuidev4.2_Final.pdf)
[NSX-User-Guide-4.2.2.pdf](https://github.com/apache/cloudstack/files/57306/NSX-User-Guide-4.2.2.pdf)

Please let me know if you need help with this

miguelaferreira commented:
@nvazquez I suggested to add the config to the test itself, but I gave it some more thought and I agree with you that the test should just assume the config is in place. 

Thanks for the documents, but it is still a lot to digest. What about step-wise instructions on what to create and in which order?

serg38 commented:
Assuming everything else is ready e.g. service controllers, transport zone, management server, STT tunnels in-between then to create L2 gateway (page 89 of the User guide):
1. Create new gateway transport node
1.1. Download and install either OVA or ISO for your hypervisor (available on Vmware site). You need at least 2 separate vSwitches. If tunneling traffic follows a different path, then 3 interface might be needed for the appliance
1.2 Login with admin/admin
1.3 Configure
set hostname <desired hostname>
add network dns-server <DNS server address>
add network ntp-server 0.pool.ntp.org
set network interface breth1 ip config static <desired-IP-address> <netmask>   .it could be breth0 in some cases and in some cases more than one bridge
add switch manager <controller-IP-address>
1.4. Obtain gateway certificate
show switch certificate 
1.5 Add gateway node to NSX
  Transport Node->Add->Gateway->Display Name->Admin Status Enabled->Security Credentials - Certificate. Paste from section 1.4->Connection->Add->STT->IP of the bridge used for tunneling traffic
1.6. Create L2 Gateway Service and publish it on created Gateway transport node

miguelaferreira commented:
Thanks for the instructions @serg38 

I will attempt to reproduce this early this week.

miguelaferreira commented on tools/marvin/marvin/lib/base.py:
```diff
@@ -3876,7 +3876,12 @@ def add(cls, apiclient, services, physicalnetworkid,
+        if l2gatewayserviceuuid:
+            cmd.l2gatewayserviceuuid = l2gatewayserviceuuid
+        else:
+            cmd.l2gatewayserviceuuid = services['l2gatewayserviceuuid']
```
There should be the possibility to create the device without the l2 gateway uuid. This construct forces one to exists. Please make sure you test of `services` has key `'l2gatewayserviceuuid'`before you retrieve it.

miguelaferreira replied:
Never mind this one, I've done it myself and I will ask you to pull in my changes to your PR asap.

miguelaferreira commented:
I've tested this PR with the additional [commits](https://github.com/nvazquez/cloudstack/pull/1) that I've requested to be merged here.

My environment:
* Management Server
* 2 KVM hypervisors
* 3 NSX controllers
* 1 NSX Service Node
* 2 NSX Gateways
* 1 NSX Manager

NSX setup:
* 1 transport zone (with service node, hypervisors and gateways configured in it)
* 1 L2 Gateway service (connected to a NSX Gateway VM)
* 1 L3 Gateway service (connected to a NSX Gateway VM)
* 1 LRouter connected to the L# Gateway service

Same [marvin configuration](https://github.com/miguelaferreira/MCT-shared/blob/9b64fd70a765e35778b1cc768edc03dbb3924654/marvin/mct-zone1-kvm1-kvm2-NVP.cfg) used for deploying DC and running tests.

Tests:
```
nosetests --with-marvin --marvin-config=/data/shared/marvin/mct-zone1-kvm1-kvm2-NVP-2.cfg -s -a tags=advanced,required_hardware=true test/integration/plugins/test_nicira_controller.py
(...)
test_01_nicira_controller (test_nicira_controller.TestNiciraContoller) ... === TestName: test_01_nicira_controller | Status : SUCCESS ===
ok
Nicira clusters will redirect clients (in this case ACS) to the master node. ... === TestName: test_02_nicira_controller_redirect | Status : SUCCESS ===
ok
test_03_nicira_tunnel_guest_network (test_nicira_controller.TestNiciraContoller) ... === TestName: test_03_nicira_tunnel_guest_network | Status : SUCCESS ===
ok
Shared Networks Support ... === TestName: test_04_nicira_shared_networks_numerical_vlanid | Status : SUCCESS ===
ok
Shared Networks Support ... === TestName: test_05_nicira_shared_networks_lrouter_uuid_vlan_id | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 5 tests in 1961.460s

OK
```

@nvazquez if you can reproduce the same test results in your environment, then I think we have tested this PR thoroughly.

miguelaferreira commented:
@DaanHoogland I can't figure out why the rats build is failing. Can you please help us?

DaanHoogland commented:
@miguelaferreira the report says api/src/com/cloud/network/guru/NetworkGuruAdditionalFunctions.java doesn't contain a known license

DaanHoogland commented on api/src/com/cloud/network/guru/NetworkGuruAdditionalFunctions.java:
```diff
@@ -0,0 +1,12 @@
+package com.cloud.network.guru;
```
a license header is missing here,

nvazquez replied:
@DaanHoogland Thanks Daan. I added it now

remibergsma commented:
Nice work guys!

@nvazquez This will be merged to master, which is now at `4.8.0-SNAPSHOT`. Can you please rebase with master, and move your SQL change to `schema-471to480.sql` (https://github.com/apache/cloudstack/blob/master/setup/db/db/schema-471to480.sql).

@miguelaferreira Can you conclude your review with a LGTY?

miguelaferreira commented:
LGTM!

nvazquez commented:
@remibergsma Thanks a lot Remi! I rebased with master, moved sql to 4.8.0 schema and pushed it again.
@miguelaferreira Thanks a lot for your help and your effort testing this features!

Merry Chistmas guys

remibergsma commented:
@DaanHoogland Jenkins failed due to this (missing artifact). Any hint on the cause?
```
Archiving artifacts
ERROR: No artifacts found that match the file pattern "target/rat.txt". Configuration error?
ERROR: ?target/rat.txt? doesn?t match anything: ?target? exists but not ?target/rat.txt?
Build step 'Archive the artifacts' changed build result to FAILURE
Skipping Cobertura coverage report as build was not UNSTABLE or better ...
Publishing Clover coverage report...
No Clover report will be published due to a Build Failure
Putting comment on the pull request
Finished: FAILURE
```

DaanHoogland commented:
@remibergsma (and others) I will continue to look at this. It has kind of a long and irregular round trip sorry for the inconvenience.

DaanHoogland commented:
btw the reason is that rat.txt was deleted if it succeeded by the next build step (mvn clean)

DaanHoogland commented:
@nvazquez If you open/close this PR it should build again at builds.a.o. The job should now be fully functional.

nvazquez commented:
@DaanHoogland Done

DaanHoogland commented:
hm, that didn't work, strange it worked for me. another option would be to force push. If you feel the extra run is needed you can try that. This looks like a well tested and reviewed PR however. I doubt if the extra run will give us any information we don't have yet.

nvazquez commented:
@DaanHoogland I see that all checks have passed, what is it failing?

DaanHoogland commented:
@nvazquez the job has not rerun, but as far ad I can see it is alright.

miguelaferreira commented:
@DaanHoogland @remibergsma (cc @nvazquez) can we merge this PR before master drifts away again?

DaanHoogland commented:
@miguelaferreira I see no issues so LGTM (no integration tests done!)

remibergsma commented:
Hi @nvazquez, when we run this in production weird stuff started to happen. This happened with VPCs that have a private gateway attached. It seemed to get confused and plugged nics in the wrong lswitches. We had to revert this change and it worked fine again. Before digging deeper, did you ever succeed creating a private gateway with this code base?

nvazquez commented:
Hi @remibergsma no, I didn't try VPC. I meant to adjust shared network handling, technically any isolated stuff including VPC would be affected. But I can help you examining this code, if you need any help please let me know
